### ISSUE TITLE
page source code

### ISSUE THREAD

**Issue by ghost** [zzzprojects/html-agility-pack#545]
hi

I want to get the tracklist from this page using html agility pack

https://www.aboveandbeyond.nu/radio/abgt484

But when I want to look at the source code of the page using my webbrowser, I get the code below.
The tracklist can't be found in the code!

Can you tell me how this is possible? The page is working fine in my browser if you click it on this page (top link)

https://www.aboveandbeyond.nu/abgt

How to use html agility pack on this page?

```
<body>
    <div class="main">
      <div class="card">
        <div class="header">
          <h1>Page Not Found</h1>
        </div>
        <div class="body">
          <p>Looks like you've followed a broken link or entered a URL that doesn't exist on this site.</p>
          <p>
            <a id="back-link" href="/">
              <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16">
                <path fill="#007067" d="M11.9998836,4.09370803 L8.55809517,7.43294953 C8.23531459,7.74611298 8.23531459,8.25388736 8.55809517,8.56693769 L12,11.9062921 L9.84187871,14 L4.24208544,8.56693751 C3.91930485,8.25388719 3.91930485,7.74611281 4.24208544,7.43294936 L9.84199531,2 L11.9998836,4.09370803 Z"/>
              </svg>
              Back to our site
             </a>
          </p>
          <hr><p>If this is your site, and you weren't expecting a 404 for this path, please visit Netlify's <a href="https://answers.netlify.com/t/support-guide-i-ve-deployed-my-site-but-i-still-see-page-not-found/125?utm_source=404page&utm_campaign=community_tracking">"page not found" support guide</a> for troubleshooting tips.
          </p>
        </div>
      </div>
    </div>
    <script>
      (function() {
        if (document.referrer && [document.location.host](https://document.location.host/) && document.referrer.match(new RegExp("^https?://" + [document.location.host](https://document.location.host/)))) {
          document.getElementById("back-link").setAttribute("href", document.referrer);
        }
      })();
    </script>
  </body>
```

**Comment by elgonzo** [comment]
> Can you tell me how this is possible?

Because the page you see in the web browser is **dynamically** created by some Javascript or WASM in your web browser.
What your web browser presents as "source code" is not the page code as provided by the web server. What the browser presents as "source code" is simply an HTML rendering of the page content (the DOM of the page), and that can include parts of the original HTML page delivered by the web server as well as page element dynamically created by Javascript.


HAP is only a HTML **parser**. It is not a browser engine, and thus does not feature a Javascript engine backed by a DOM. So, processing the HTML page from the web server in HAP, you will only get that and no dynamically created page content...

**Comment by JonathanMagnan** [comment]
In addition to @elgonzo answer,

It's somewhat possible to get a dynamic text through HAP if you still use the .NET Framework (require a `WebBroswer): https://html-agility-pack.net/from-browser

However, I highly recommend Selenium. Here is a short tutorial about [Selenium WebDriver](https://riptutorial.com/selenium-webdriver). Learning it will allow you to push everything further, such as if you need something that logs automatically into a website and grabs some text after.

Best Regards,

Jon

**Comment by elgonzo** [comment]
_TL;DR: Just don't bother with HTML parsing in this case. Json ftw!_

Since i was bored (boredom be praised!), i took a look at what data in which form the page gets and processes, using the network monitor of the web development tools of my web browser. Doing that often gets you a lot of insight into how a web site operates, which in turn can (and in this case does) make your life much easier.

Basically, all the dynamic page content is generated based on **json data** fetched from the web sever. And it should be much easier and less cumbersome to process the json data using System.Text.Json compared to trying to extract the same information from some HTML structures...

The list of the shows is fetched from https://www.aboveandbeyond.nu/api/abgt/abgt.json. Each show entry in this json features an "identifier". That identifier value is being used to create the URL for the json data with the track list of the respective show based on the URL template `https://www.aboveandbeyond.nu/api/abgt/<identifier>.json`. So, if the show identifier is for example abgt473, the URL with the track list json would be https://www.aboveandbeyond.nu/api/abgt/abgt473.json.

**Comment by elgonzo** [comment]
> Can you give me an example for reading the JSON in .NET framework? 

Depends on which .NET version you are targetting. If you are using .NET 5 or newer, use System.Text.Json (which is included in modern .NET versions). If you are still on an old .NET Framework 4.x build target i would strongly recommend to migrate to .NET 8, unless there are road-blocks preventing you from doing that. Otherwise, if your build target is .NET Framework 4.6 or newer, you can still use System.Text.Json, but you need to import it into your project as a nuget package (https://www.nuget.org/packages/System.Text.Json/8.0.3#supportedframeworks-body-tab).

If your build target is older than .NET Framework 4.6, run for the hills ;-P The only practical choice for a Json processor/serializer on such old .NET frameworks is Json.NET (which is also often called Newtonsoft.Json).

Regardless of whether you use STJ or Json.NET/Newtonsoft.Json, you have two approaches available to consume the json data:

1. Create model classes with properties that match the logical data structures and data types of the values in the fetched json _exactly. The json (de)serializer is then practically able to automatically create instances of these model classes with the values from the json data. Easy peasy, the only thing you need to be careful with is that your model classes _really exactly match_ the json data layout and value types.

2. Consume data from the json documents without deserializing the json into instances of model classes. Both STJ and Json.NET allow you to do that relatively easily by providing types that represent the data structure and values in a json document directly..

(There is also a third approach by reading the json token stream sequentially, requiring you to track nested data structures in the json data yourself. But that doesn't make any sense for what you are trying to do so just ignore that third possible option for reasons of impracticality.)


There should be tutorials about either approach with STJ or Json.NET/Newtonsoft.Json out there -- the world is at your finger tips. If you can't get a handle on how to do that correctly, i'd suggest to seek help on sites like stackoverflow.com. In case you are unfamiliar with stackoverflow.com: Please make an research effort before asking a question, and pay attention to _how_ you ask and that you format the question in a readable manner presenting relevant information pertinent to your problem. The people helping there are unpaid volunteers, and they often unceremoniously close questions if the question gives the impression the asker wants to delegate their work and not bother with themselves putting effort into their work nor the question. To be clear, i am not trying to insinuate that you do, but i want to emphasize that your chances of getting a good answer there often depends to a high degree on how you ask, how you format the question and that you [stick to their rules if you ask on stackoverflow.com](https://stackoverflow.com/help/how-to-ask).

**Comment by elgonzo** [comment]
> I was just trying but using the example from https://html-agility-pack.net/from-browser , I get this error even in .net framework 4.8: `Error CS1061 'HtmlWeb' does not contain a definition for 'LoadFromBrowser' and no accessible extension method 'LoadFromBrowser' accepting a first argument of type 'HtmlWeb' could be found (are you missing a using directive or an assembly reference?)`

Regardless of that error, i wouldn't know how `HtmlWeb.LoadFromBrowser` could help you in your particular case. Remember, the dynamic HTML with the track list is created by a Javascript click event handler. Therefore, to materialize the HTML with the track list, you have to to trigger/call this click event handler somehow. I mean, how would you do this with `HtmlWeb.LoadFromBrowser`?

**Comment by elgonzo** [comment]
My apologies, but i am not going to tutor you here, as this is the issue tracker for HAP and therefore not the right place for that. My comments re json were just meant to point out a more direct way to get the data without needing to jump through the hoops of capturing dynamically created HTML and parsing this HTML for information that is readily available in a different format (json).

I suggested to you to seek out tutorials and/or ask question(s) on stackoverflow.com, and giving you pointers regarding the available general approaches with both STJ and  Json.NET/Newtonsoft.Json, so you can do a rough initial calibration of which approach feels better to you and therefore what kind of tutorials to seek and prefer. I also suggested to ask on stackoverflow.com. It's entirely up to you what you take from my suggestions...

**Comment by JonathanMagnan** [comment]
> Therefore, to materialize the HTML with the track list, you have to to trigger/call this click event handler somehow. I mean, how would you do this with `HtmlWeb.LoadFromBrowser`?

The `LoadFromBwoser` uses really a web browser, so technically, you can do everything. I will never recommend this solution over using Selenium web browser, but here is an example of clicking on the `Download` button on our page and loading the source of the download page instead of the home:

```csharp
var url = "https://html-agility-pack.net/";

var web1 = new HtmlAgilityPack.HtmlWeb();
var isLinkClicked = false;

var doc1 = web1.LoadFromBrowser(url, o =>
{
	var webBrowser = (WebBrowser)o;

	if(!isLinkClicked)
	{
		HtmlElementCollection links = webBrowser.Document.GetElementsByTagName("a");

		foreach (HtmlElement link in links)
		{
			if (link.InnerText == "  Download   ")
			{
				link.InvokeMember("click");
				isLinkClicked = true;
				Thread.Sleep(2000);
				break; // Stop the loop after clicking the right element
			}
		}
	}

	// WAIT until the link has been clicked and an element you want on the new page is now accessible
	return isLinkClicked && title == "Download HTML Agility Pack \r\nHTML parser to read/write DOM\r\n ";
});

// HAP have loaded the source of the `/download/` page, not the home
var text = doc1.DocumentNode.InnerHtml;
```

---

Out of topic, but Above and Beyond was amazing when I saw them a few month ago in Quebec ;)

**Comment by JonathanMagnan** [comment]
Hello @trance-babe ,

Could you provide the reference path of your `HtmlAgilityPack` dll such as: `C:\Repos\HtmlAgilityPack\packages\HtmlAgilityPack.1.11.60\lib\Net45\HtmlAgilityPack.dll`

One reason that could explain why you don't see it is if you point on a .NET Standard version instead of the version Net45 like me.

Best Regards,

Jon

**Comment by JonathanMagnan** [comment]
Here is an example:

![image](https://github.com/zzzprojects/html-agility-pack/assets/23240780/5d708396-579d-497b-98e3-b098387fb21b)

**Comment by JonathanMagnan** [comment]
You currently provided the path of `HtmlAgilityPack.CssSelectors` library, not the `HtmlAgilityPack`.

Could you show the path of `HtmlAgilityPack` itself?
